rio_tiler/io/base.py

```diff
@@ -459,31 +459,18 @@ def tile(
 
         def _reader(asset: str, *args: Any, **kwargs: Any) -> ImageData:
             url = self._get_asset_url(asset)
+            expr = asset_expression.get(asset)  # type: ignore
+            idx = asset_indexes.get(asset) or kwargs.pop("indexes", None)  # type: ignore
             with self.reader(url, tms=self.tms, **self.reader_options) as cog:  # type: ignore
-                data = cog.tile(
-                    *args,
-                    indexes=asset_indexes.get(asset, kwargs.pop("indexes", None)),  # type: ignore
-                    expression=asset_expression.get(asset),  # type: ignore
-                    **kwargs,
-                )
+                data = cog.tile(*args, indexes=idx, expression=expr, **kwargs)
                 data.band_names = [f"{asset}_{n}" for n in data.band_names]
                 return data
 
-        output = multi_arrays(
-            assets,
-            _reader,
-            tile_x,
-            tile_y,
-            tile_z,
-            **kwargs,
-        )
-
+        img = multi_arrays(assets, _reader, tile_x, tile_y, tile_z, **kwargs)
         if expression:
-            blocks = get_expression_blocks(expression)
-            output.data = apply_expression(blocks, assets, output.data)
-            output.band_names = blocks
+            img.apply_expression(expression, bands=assets)  # type: ignore
```

Comment:
here we overwrite the `bands` that will be use as expression.

FYI:
- we now have `band_names = ["green_b1", "red_b1"]`
- up to now we are assuming that `expression=green/red` mean there is only one band in each dataset **THIS IS BAD**
- in this line we tell `apply_expression` to overwrite the bands_names to `["green", "red"]` to keep the compatibility.
- The proper expression should be `expression="green_b1/red_b1"` but this will be maybe too big of a **breaking change**

@geospatial-jeff any advice here?

Reply:
opened #524 to talk about this

Reply:
> The proper expression should be expression="green_b1/red_b1" but this will be maybe too big of a breaking change

I think we should do something like this.  We are breaking things anyways so might as well get it right now.  Representing the full hierarchy of `asset::band` seems like the best way to support applying expressions across multi-band assets.